### PR TITLE
Various (many...) interface and Reporter enhancements

### LINKED ISSUE #17605: Improve Reporter error handling, simplify get/declare
## Reason
The Reporter system could benefit from a variety of enhancements.

## Design
- Declare Reporter values for Postprocessors and VectorPostprocessors in the objects themselves, removing the need for things like SetupPostprocessorDataAction
- Store producer and consumer MooseObjects in the ReporterState itself so that we can report better errors when things weren't produced, or things were requested/declared with the same name and same type
- Keep track of the state when a Reporter is actually declared a Postprocessor or VectorPostprocessor, both for output and error handling
- Add checking on type when declaring/requesting Reporters

## Impact
Will enhance public APIs and add more usable errors. Should not effect the use of the system.

### PR DESCRIPTION
**Pull request by loganharbour**
Most significant changes in PostprocessorInterface, VectorPostprocessorInterface, and UserObjectInterface:

- Makes various methods const
- Improves error checking; uses paramError where possible
- Restricts the checking of objects exist to after they are constructed
- Removes postprocessor metadata and instead does default value checking on-the-fly
- Adds dependency helpers to remove virtual overrides
- Adds error checking for all of the tests

This is ~1000 lines of tests/test objects, so it's _almost_ a reduction in framework code with lots of added functionality.

Refs #17512
Closes #17605

### CODE REVIEW

File: framework/include/ics/InitialConditionBase.h
Comment (aeslaughter) [inline framework/include/ics/InitialConditionBase.h:125]: I think this is the proper approach, but would add a detailed comment about why this is mutable.

File: framework/include/interfaces/PostprocessorInterface.h
Comment (aeslaughter) [inline framework/include/interfaces/PostprocessorInterface.h:102]: Should you make the index const here as well?

File: framework/include/problems/FEProblemBase.h
Comment (aeslaughter) [inline framework/include/problems/FEProblemBase.h:856]: Can we add some doco here, at least pointing to the ReporterData::getReporterData method. It took me a bit to remember that the "object_name" is the object asking and is simply for error reporting.

File: framework/include/reporters/ReporterData.h
Comment (aeslaughter) [inline framework/include/reporters/ReporterData.h:294]: Can you document the need for this as well.

File: framework/include/actions/ActionFactory.h
Comment (loganharbour) [reply framework/include/actions/ActionFactory.h:131]: Not at all. The std::size_t cast -> bool is in the standard, and is well defined

File: framework/src/transfers/ReporterTransferInterface.C
Comment (zachmprince) [inline framework/src/transfers/ReporterTransferInterface.C:164]: Could you just call the previous check function twice instead of having this one?

File: framework/include/outputs/ReporterDebugOutput.h
Comment (zachmprince) [inline framework/include/outputs/ReporterDebugOutput.h:27]: This comment isn't true

**Review by zachmprince — CHANGES_REQUESTED** [review]
Just some easy stuff I found. Looks really good to me.

File: framework/include/reporters/ReporterContext.h
Comment (loganharbour) [reply framework/include/reporters/ReporterContext.h:262]: Ah, yep... this should be pure virtual

File: framework/include/restart/RestartableData.h
Comment (loganharbour) [reply framework/include/restart/RestartableData.h:149]: Oops. good catch

File: framework/src/transfers/ReporterTransferInterface.C
Comment (loganharbour) [reply framework/src/transfers/ReporterTransferInterface.C:164]: 👍🏼

File: framework/include/interfaces/PostprocessorInterface.h
Comment (aeslaughter) [inline framework/include/interfaces/PostprocessorInterface.h:105]: Was dropping the `const &` intentional?

File: framework/include/reporters/ReporterContext.h
Comment (aeslaughter) [inline framework/include/reporters/ReporterContext.h:399]: Why can't this just be in ReporterContext?

File: framework/src/postprocessors/Receiver.C
Comment (aeslaughter) [inline framework/src/postprocessors/Receiver.C:32]: Why this change, seems like it should use the interface if it can.

File: framework/src/reporters/ReporterData.C
Comment (aeslaughter) [inline framework/src/reporters/ReporterData.C:33]: Where does this happen now? I am sad you are getting rid of my lambdas, I love these little guys.

**Review by aeslaughter — DISMISSED** [review]

File: framework/include/interfaces/PostprocessorInterface.h
Comment (loganharbour) [reply framework/include/interfaces/PostprocessorInterface.h:105]: Yeah. This isn't a reference for people to store. If they want the value at all, they should just call getPostprocessorValue(), which will create something they can bind to if it's a default value. I think that I should actually just remove this method. I don't think it's used.

File: framework/include/reporters/ReporterContext.h
Comment (loganharbour) [reply framework/include/reporters/ReporterContext.h:399]: Because `this` changes depending on what context you have. If you put this just in `ReporterContext`, `this` will always be of type `ReporterContext<T>`, even if you call `ReporterGeneralContext<T>.contextType()`.

File: framework/include/interfaces/PostprocessorInterface.h
Comment (aeslaughter) [reply framework/include/interfaces/PostprocessorInterface.h:105]: I say remove it if you can.

File: framework/src/postprocessors/Receiver.C
Comment (loganharbour) [reply framework/src/postprocessors/Receiver.C:32]: Yeah, I think you're right. Does the remaining stuff make sense, though? Needing write access to the data in order to request the older values. I'm trying to slowly wean off the use of the FEProblemBase calls (FEProblemBase::{getUserObject, getReporterValue, etc})

File: framework/src/reporters/ReporterData.C
Comment (loganharbour) [reply framework/src/reporters/ReporterData.C:33]: I think this should actually be an assertion now (didn't mean to remove it). It should be checked before hand in https://github.com/loganharbour/moose/blob/moar_moar_const/framework/src/reporters/ReporterData.C#L78

File: framework/include/reporters/ReporterContext.h
Comment (aeslaughter) [reply framework/include/reporters/ReporterContext.h:399]: Ah yes, stupid C++. What about making this an input to the Constructor instead? It just seems silly to have the same method in each. I am thinking something like.

```
Context(..., const std::string & context_type = ""

NewContext : Context(..., MooseUtils::prettyCppType<NewContext>());

```

**Comment by loganharbour** [comment]
The producer used to be owned by the ReporterState. It's now owned by the ReporterContext, which means that it's _required_ any time a context is defined. This makes more sense.

File: framework/include/reporters/ReporterContext.h
Comment (loganharbour) [reply framework/include/reporters/ReporterContext.h:416]: This was how I had it originally. I'm fine with that. Will move

Comment (loganharbour) [reply framework/include/reporters/ReporterContext.h:416]: Actually... this ends up adding quite a bit more code compared to just having the virtuals. I'm at a loss

File: framework/src/reporters/ReporterData.C
Comment (aeslaughter) [inline framework/src/reporters/ReporterData.C:138]: All that virtual stuff is all for this one method. Can we just used the C++ type stuff for this one piece of information?

**Review by aeslaughter — DISMISSED** [review]

**Review by aeslaughter — DISMISSED** [review]

**Review by aeslaughter — DISMISSED** [review]

**Comment by loganharbour** [comment]
@aeslaughter Need yet another review due to a conflict after rebase. Hopefully can get this in today.

**Review by aeslaughter — APPROVED** [review]

**Review by rwcarlsen — APPROVED** [review]

**Review by zachmprince — APPROVED** [review]

**Pull request merged**
